numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -1028,7 +1060,22 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
 NPY_NO_EXPORT int
 PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals)
 {
-    return PyArray_IntpFromIndexSequence(seq, vals, (npy_intp)maxvals);
+    if (PyLong_CheckExact(seq) || !PySequence_Check(seq)) {
+        vals[0] = intp_from_scalar(seq);
+        if (PyErr_Occurred() != NULL) {
+            return -1;
+        } else {
+            return 1;
+        }
+    }
+
+    PyObject *seq_obj = PySequence_Fast(seq,
+            "Expected an integer or an iterable/sequence of integers.");
+    if (seq_obj == NULL) {
+        return -1;
+    }
```

Comment:
Should we invert this here as well?  To make it "try/except style"?

Reply:
I wasn't sure since in `PyArray_IntpFromSequence` we do not need to reserve memory "manually". I'm going to refactor this as well

Reply:
Fixed in 952871653a474c61b81a286dc6797da75ca226e7, I also introduced two tests with an overflow error for more extensive checking.